### PR TITLE
Implementation/47147 extend existing view for adding storages to projects

### PR DESCRIPTION
**Pull request by akabiru**
### [Description](https://community.openproject.org/projects/openproject/work_packages/47147/activity)

Allow users to configure file storage modes for a given provider, specifying a default project folder location or custom project folder location. The configured project folder location is where file uploads will be done by default.

### Tasks

- [x] Include the `project_folder_mode` in the add storage form
   -   ~Handle case for existing storages (set a default / leave unset?)~
   - [x] Add raw string selection for the `project_folder_id` when `manual` selection as a temp solution pending angular component selection
   - [x] test edge case: can only save `project_folder_id` when `manual` project folder mode
   - [x] test edge case: editing from manual to inactive should retain previous value
   - ~Add view component unit tests~ covered by feature specs
   - ~Look into controller specs~
- [x] Add edit workflow
- [x] UI
   - [x] Update Edit title
   - [x] Update Add & Edit Button
   - [x] Add Cancel button, redirects back to index page
   - [x] In Edit mode, render provider as text
   - [x] Update storage admin select in edit to be consistent with file storages
- [x] Wrap functionality behind a feature flag
    - [x] `OpenProject::FeatureDecisions.add :storage_project_folders`
    - [x] Hide radio buttons + text field
    - [x] Hide Edit button

### Screenshots

<details> <summary> Click Here 🔽  </summary>

Screenshot 2023-05-03 at 11 26 10 AM
Screenshot 2023-05-03 at 11 26 01 AM
Screenshot 2023-05-03 at 11 25 43 AM
Screenshot 2023-05-03 at 11 25 32 AM
Screenshot 2023-05-03 at 11 24 52 AM

</details>

### CODE REVIEW

File: modules/storages/app/views/storages/admin/storages/_form.html.erb [hunk 1]
```diff
@@ -33,17 +33,25 @@ See COPYRIGHT and LICENSE files for more details.
...
+      %>
+      <span class="form--field-instructions"><%= t('storages.instructions.type') %> <%= t('storages.instructions.type_link_text') %></span>
+    <% when 'edit' %>
+      <%= f.label :provider_type, class: "form--label" do %>
+        <%= t(:"storages.label_provider_type") %>
+      <% end %>
+      <div class="form--field-container -horizontal">
+        <span><%= @object.name %></span>
+      </div>
+    <% end %>
```
Comment (Kharonus) [inline modules/storages/app/views/storages/admin/storages/_form.html.erb:54]: Was this discussed, that we disable switching the storage type when editing a storage?

File: modules/storages/app/views/storages/project_settings/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+      $("input[type='radio']").change(function() {
+        if ($(this).val() == 'manual') {
+          $("#storages_manual_folder_selection").show();
+        } else {
+          $("#storages_manual_folder_selection").hide();
+        }
+      })
+    });
+  <% end -%>
```
Comment (Kharonus) [inline modules/storages/app/views/storages/project_settings/_form.html.erb]: Did you talk to Oliver about the dynamic scripts? There are IMHO two things, I would suggest as an improvement:

- use the meta tag for required scripts

```erb
<% content_for :header_tags do %>
  <meta name="required_script" content="project_folder_mode_selection" />
<% end %>
```

- get rid of jQuery and write a basic typescript file

If there is a good reason not to do this right now, we should do this in the next step, when this script will get much more complex.

**Review by Kharonus — COMMENTED** [review]
Looks very good in general!

I'd like to raise the questions about the dynamic script, written below.

File: modules/storages/app/views/storages/admin/storages/_form.html.erb (hunk 1 above)
Comment (akabiru) [reply modules/storages/app/views/storages/admin/storages/_form.html.erb:54]: Yes, we discussed this with @wielinde and thought to make it consitent with the project settings form. Do you foresee any drawbacks with this approach? I would presume editing the storage provider after the fact would be problematic either way?

File: modules/storages/app/views/storages/project_settings/_form.html.erb (hunk 2 above)
Comment (akabiru) [reply modules/storages/app/views/storages/project_settings/_form.html.erb]: Good shout! As discussed, we'll pair to address the above. 👍🏾

File: modules/storages/app/views/storages/admin/storages/_form.html.erb (hunk 1 above)
Comment (wielinde) [reply modules/storages/app/views/storages/admin/storages/_form.html.erb:54]: Yes, I think we three discussed that together. Switching the storage type in edit mode would only make sense if all or most of the storage attributes would stay the same. This is not the case here. Most of the fields would need to get reset once the storage type changes. Then I believe it is better to tell the user to start a storage from scratch.
